MAINT Introduce `BaseCriterion` as a base abstraction for `Criterion`s

adam2392 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
Fixes #24577 


#### What does this implement/fix? Explain your changes.
Adds the `BaseCriterion` class as an abstract class for just the API relevant changes, while keeping `Criterion` as the assumed "supervised learning" criterion.

This change is backwards compatible.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

ogrisel commented:
I think this PR includes by mistake unrelated Cython files from the `sklearn/metrics` folder.

ogrisel reviewed (commented):
You might want to check for `BaseCriterion` instead of `Criterion` in:

- https://github.com/scikit-learn/scikit-learn/blob/bcff21e0ed0a87313432b669bbe0f8c01fa1f35d/sklearn/tree/_classes.py#L324

to make it easier to subclass `BaseDecisionTree` with a custom criterion although other parts of the `fit` method assume a float64 dtyped `y` array. The various steps of the fit method should probably be split in private submethod to make it easier to subclass. This can be addressed in subsequent PRs.

glemaitre commented on sklearn/metrics/_pairwise_distances_reduction/_radius_neighborhood.pxd:
```diff
@@ -0,0 +1,152 @@
+cimport numpy as cnp
```
The pairwise part has been added by mistake I think.

adam2392 replied:
Ooo yeah not sure why rebasing got this. I'll just redo the changes from main.

adam2392 commented on sklearn/tree/_criterion.pxd:
```diff
@@ -15,7 +15,30 @@ from ._tree cimport SIZE_t           # Type for indices and counters
+    cdef double weighted_n_right         # Weighted number of samples in the right node
+
+    # The criterion object is maintained such that left and right collected
+    # statistics correspond to samples[start:pos] and samples[pos:end].
```
True I modified the subclassing structure

adam2392 committed:
Fix rebase

adam2392 committed:
Fix clean

adam2392 committed:
Changing criteiron to basecriterion

adam2392 committed:
Run black on sklearn tree classes

adam2392 committed:
Add changelog

adam2392 committed:
Fix spacing

adam2392 committed:
Fix changelog

adam2392 commented on sklearn/tree/_criterion.pyx:
```diff
@@ -192,6 +170,40 @@ cdef class Criterion:
+    """Interface for impurity criteria.
+
+    This object stores methods on how to calculate how good a split is using
+    different metrics.
```
Good point. Added that comment in and some more description overall.

adam2392 commented:
Cross-linking: https://github.com/scikit-learn/scikit-learn/discussions/20648

adam2392 commented:
Hi @jjerphan @ogrisel and @glemaitre just wanted to do a light ping to see if there's any feedback regarding this and if this is good to merge? This is probably the most simple change I proposed in the roadmap for improving trees.

jjerphan commented on sklearn/tree/_criterion.pyx:
```diff
@@ -192,6 +170,41 @@ cdef class Criterion:
+cdef class Criterion(BaseCriterion):
+    """Interface for impurity criteria.
+
+    This object stores methods on how to calculate how good a split is using
+    different metrics. This is the base class for any supervised tree criterion
+    model with a homogeneous float64 dtyped y.
```
```suggestion
    Instances of this class are responsible for compute splits' impurity difference

    Criterion is the base class for criteria used in supervised tree-based models
    with a homogeneous float64-dtyped y.
```

jjerphan commented on sklearn/tree/_criterion.pyx:
```diff
@@ -192,6 +170,41 @@ cdef class Criterion:
+        Parameters
+        ----------
+        y : array-like, dtype=DOUBLE_t
+            y is a buffer that can store values for n_outputs target variables
```
```suggestion
            y is a buffer that can store values for the `n_outputs` target variables.
```

jjerphan commented on sklearn/tree/_criterion.pyx:
```diff
@@ -192,6 +170,41 @@ cdef class Criterion:
+        sample_weight : array-like, dtype=DOUBLE_t
+            The weight of each sample
+        weighted_n_samples : double
+            The total weight of the samples being considered
```
```suggestion
            The sum of the weights of the samples being considered.
```

jjerphan replied:
```suggestion
            The sum of the weights of the samples being considered.
```

jjerphan replied:
```suggestion
            The sum of the weights of the samples being considered.
```

jjerphan commented on sklearn/tree/_criterion.pyx:
```diff
@@ -192,6 +170,41 @@ cdef class Criterion:
+        ----------
+        y : array-like, dtype=DOUBLE_t
+            y is a buffer that can store values for n_outputs target variables
+        sample_weight : array-like, dtype=DOUBLE_t
+            The weight of each sample
```
```suggestion
        sample_weight : pointer to a buffer of DOUBLE_t
            The pointer to the buffer storing each sample weight.
```

jjerphan replied:
```suggestion
        sample_weight : pointer to a buffer of DOUBLE_t
            The pointer to the buffer storing each sample weight.
```

jjerphan commented on sklearn/tree/_criterion.pyx:
```diff
@@ -192,6 +170,41 @@ cdef class Criterion:
+        samples : array-like, dtype=SIZE_t
+            Indices of the samples in X and y, where samples[start:end]
+            correspond to the samples in this node
+        start : SIZE_t
+            The first sample to be used on this node
+        end : SIZE_t
+            The last sample used on this node
```
```suggestion
        start : SIZE_t
            The index of first sample in `samples` to be considered in this node.
        end : SIZE_t
            The index of  last sample in `samples` to be considered in this node.
```

jjerphan commented on sklearn/tree/_criterion.pxd:
```diff
@@ -52,10 +52,24 @@ cdef class Criterion:
+    cdef const DOUBLE_t[:, ::1] y        # Values of y
+
+    # Methods
+    cdef int init(self, const DOUBLE_t[:, ::1] y, DOUBLE_t* sample_weight,
+                  double weighted_n_samples, SIZE_t* samples, SIZE_t start,
+                  SIZE_t end) nogil except -1
```
Also, it might be possible to use `const DOUBLE_t[:]` for `sample_weights`.

adam2392 replied:
Making a note here for a future issue:

To explore changing this to a memview instead of a pointer, see https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html.

What you will need to do is change it here in the function header, in the function instantiation and then ctrl+f anything that "calls" sample_weight. However, imo after ctrl+fing in the tree/ submodule, this would be a larger change that I would be in favor of doing in a separate PR given PRing into scikit-learn is already pretty challenging.

Original comment: https://github.com/scikit-learn/scikit-learn/pull/24678/files#r1013113564

adam2392 replied:
Made an issue.

jjerphan reviewed (commented):
Thank you for this work and for the heads-up, @adam2392.

Is it possible to run the RandomForest ASV benchmarks to assess any potential performance regression w.r.t `main`?

In the meantime, here are some comments and suggestions.

adam2392 commented:
Looks like there are no performance regressions:

```
(sklearn) adam2392@Adams-MacBook-Pro asv_benchmarks % asv show crit
Commit: crit <crit>

ensemble.RandomForestClassifierBenchmark.peakmem_fit [adam2392_mbp/conda-py3.9-cython-joblib-numpy-pandas-scipy-threadpoolctl]
  ok
  ================ ======
  --               n_jobs
  ---------------- ------
   representation    1   
  ================ ======
       dense        248M 
       sparse       506M 
  ================ ======
  started: 2022-11-08 11:30:23, duration: 12.9s

ensemble.RandomForestClassifierBenchmark.peakmem_predict [adam2392_mbp/conda-py3.9-cython-joblib-numpy-pandas-scipy-threadpoolctl]
  ok
  ================ ======
  --               n_jobs
  ---------------- ------
   representation    1   
  ================ ======
       dense        220M 
       sparse       442M 
  ================ ======
  started: 2022-11-08 11:30:36, duration: 2.03s

ensemble.RandomForestClassifierBenchmark.time_fit [adam2392_mbp/conda-py3.9-cython-joblib-numpy-pandas-scipy-threadpoolctl]
  ok
  ================ ============
  --                  n_jobs   
  ---------------- ------------
   representation       1      
  ================ ============
       dense        4.57±0.03s 
       sparse       7.25±0.01s 
  ================ ============
  started: 2022-11-08 11:30:38, duration: 1.16m

ensemble.RandomForestClassifierBenchmark.time_predict [adam2392_mbp/conda-py3.9-cython-joblib-numpy-pandas-scipy-threadpoolctl]
  ok
  ================ ==========
  --                 n_jobs  
  ---------------- ----------
   representation      1     
  ================ ==========
       dense        141±4ms  
       sparse       948±10ms 
  ================ ==========
  started: 2022-11-08 11:31:47, duration: 50.9s

ensemble.RandomForestClassifierBenchmark.track_test_score [adam2392_mbp/conda-py3.9-cython-joblib-numpy-pandas-scipy-threadpoolctl]
  ok
  ================ ====================
  --                      n_jobs       
  ---------------- --------------------
   representation           1          
  ================ ====================
       dense        0.7502017055240395 
       sparse       0.8656423941766682 
  ================ ====================
  started: 2022-11-08 11:32:38, duration: 1.24s

ensemble.RandomForestClassifierBenchmark.track_train_score [adam2392_mbp/conda-py3.9-cython-joblib-numpy-pandas-scipy-threadpoolctl]
  ok
  ================ ====================
  --                      n_jobs       
  ---------------- --------------------
   representation           1          
  ================ ====================
       dense        0.9971484595841501 
       sparse       0.9996123288718864 
  ================ ====================
  started: 2022-11-08 11:32:40, duration: 1.99s
```

```(sklearn) adam2392@Adams-MacBook-Pro asv_benchmarks % asv show main
Commit: main <main>

ensemble.RandomForestClassifierBenchmark.peakmem_fit [adam2392_mbp/conda-py3.9-cython-joblib-numpy-pandas-scipy-threadpoolctl]
  ok
  ================ ======
  --               n_jobs
  ---------------- ------
   representation    1   
  ================ ======
       dense        255M 
       sparse       450M 
  ================ ======
  started: 2022-11-08 11:33:00, duration: 13.4s

ensemble.RandomForestClassifierBenchmark.peakmem_predict [adam2392_mbp/conda-py3.9-cython-joblib-numpy-pandas-scipy-threadpoolctl]
  ok
  ================ ======
  --               n_jobs
  ---------------- ------
   representation    1   
  ================ ======
       dense        218M 
       sparse       432M 
  ================ ======
  started: 2022-11-08 11:33:13, duration: 2.20s

ensemble.RandomForestClassifierBenchmark.time_fit [adam2392_mbp/conda-py3.9-cython-joblib-numpy-pandas-scipy-threadpoolctl]
  ok
  ================ ============
  --                  n_jobs   
  ---------------- ------------
   representation       1      
  ================ ============
       dense        4.72±0.04s 
       sparse       7.30±0.01s 
  ================ ============
  started: 2022-11-08 11:33:16, duration: 1.18m

ensemble.RandomForestClassifierBenchmark.time_predict [adam2392_mbp/conda-py3.9-cython-joblib-numpy-pandas-scipy-threadpoolctl]
  ok
  ================ ==========
  --                 n_jobs  
  ---------------- ----------
   representation      1     
  ================ ==========
       dense        144±10ms 
       sparse       960±30ms 
  ================ ==========
  started: 2022-11-08 11:34:27, duration: 51.2s

ensemble.RandomForestClassifierBenchmark.track_test_score [adam2392_mbp/conda-py3.9-cython-joblib-numpy-pandas-scipy-threadpoolctl]
  ok
  ================ ====================
  --                      n_jobs       
  ---------------- --------------------
   representation           1          
  ================ ====================
       dense        0.7502017055240395 
       sparse       0.8656423941766682 
  ================ ====================
  started: 2022-11-08 11:35:18, duration: 1.24s

ensemble.RandomForestClassifierBenchmark.track_train_score [adam2392_mbp/conda-py3.9-cython-joblib-numpy-pandas-scipy-threadpoolctl]
  ok
  ================ ====================
  --                      n_jobs       
  ---------------- --------------------
   representation           1          
  ================ ====================
       dense        0.9971484595841501 
       sparse       0.9996123288718864 
  ================ ====================
  started: 2022-11-08 11:35:19, duration: 2.00s
```

adam2392 commented:
@jshinm has volunteered to try to address the comments you outlined in the code. Thanks!

adam2392 commented:
I re-ran the asv benchmark after closing all open apps on my Mac M1:

```
(sklearn) adam2392@Adams-MacBook-Pro asv_benchmarks % asv compare --split upstream/main crit

Benchmarks that have improved:

       before           after         ratio
     [3e47fa91]       [106b01cf]
     <main>           <crit>    
-            555M             495M     0.89  ensemble.RandomForestClassifierBenchmark.peakmem_fit('sparse', 1)

Benchmarks that have stayed the same:

       before           after         ratio
     [3e47fa91]       [106b01cf]
     <main>           <crit>    
             267M             281M     1.05  ensemble.RandomForestClassifierBenchmark.peakmem_fit('dense', 1)
             218M             218M     1.00  ensemble.RandomForestClassifierBenchmark.peakmem_predict('dense', 1)
             439M             435M     0.99  ensemble.RandomForestClassifierBenchmark.peakmem_predict('sparse', 1)
          4.82±0s          4.75±0s     0.98  ensemble.RandomForestClassifierBenchmark.time_fit('dense', 1)
       7.25±0.01s          7.13±0s     0.98  ensemble.RandomForestClassifierBenchmark.time_fit('sparse', 1)
          137±8ms        132±0.4ms     0.97  ensemble.RandomForestClassifierBenchmark.time_predict('dense', 1)
         980±40ms         907±50ms     0.93  ensemble.RandomForestClassifierBenchmark.time_predict('sparse', 1)
  0.7502017055240395  0.7502017055240395     1.00  ensemble.RandomForestClassifierBenchmark.track_test_score('dense', 1)
  0.8656423941766682  0.8656423941766682     1.00  ensemble.RandomForestClassifierBenchmark.track_test_score('sparse', 1)
  0.9971484595841501  0.9971484595841501     1.00  ensemble.RandomForestClassifierBenchmark.track_train_score('dense', 1)
  0.9996123288718864  0.9996123288718864     1.00  ensemble.RandomForestClassifierBenchmark.track_train_score('sparse', 1)
```

There are no performance regressions with including this refactoring. However, not sure why we saw an improvement in memory for sparse data. Might be chance? Saw similar results when I ran it yesterday tho (https://github.com/scikit-learn/scikit-learn/pull/24678#issuecomment-1307752382) cc: @jjerphan 

Our next steps are for @jshinm to re-run the asv benchmarks and to also address the code changes.

jshinm committed:
remove entry from update log

jshinm committed:
move pxd comments to pyx docstring and made changes suggested by @jjerphan

jshinm committed:
move pxd comments to class docstring and make changes suggested by @jjerphan

jshinm committed:
update docstring as suggested by @jjerphan

jshinm committed:
Update sklearn/tree/_criterion.pyx

Co-authored-by: Adam Li <adam2392@gmail.com>

jshinm committed:
fix indentation as @adam2392 suggested

jshinm committed:
add contributors

adam2392 committed:
Merge pull request #27 from jshinm/jms-update-crit

adam2392 committed:
Merge branch 'main' into crit

adam2392 commented on sklearn/tree/_criterion.pyx:
```diff
@@ -29,45 +29,23 @@ from ._utils cimport WeightedMedianCalculator
+    """Abstract interface for any criterion.
 
     This object stores methods on how to calculate how good a split is using
-    different metrics.
+    a set API. 
```
Done.

adam2392 committed:
Add to the docstring

adam2392 commented:
@jjerphan this is ready now for a final review + merge.

@jshinm addressed some issues in the code review and re-ran the asv benchmark as well:

```
(crit) jshinm@jshinm-OMEN-by-HP-Laptop-16-b0xxx:~/Desktop/workstation/sklearn-jms/asv_benchmarks$ asv compare main crit

All benchmarks:

       before           after         ratio
     [9c9c8582]       [106b01cf]
     <main>           <crit>   
     
             188M             188M     1.00  ensemble.RandomForestClassifierBenchmark.peakmem_fit('dense', 1)
             422M             422M     1.00  ensemble.RandomForestClassifierBenchmark.peakmem_fit('sparse', 1)
             190M             189M     1.00  ensemble.RandomForestClassifierBenchmark.peakmem_predict('dense', 1)
             406M             406M     1.00  ensemble.RandomForestClassifierBenchmark.peakmem_predict('sparse', 1)
          3.94±0s       4.02±0.02s     1.02  ensemble.RandomForestClassifierBenchmark.time_fit('dense', 1)
       6.41±0.02s       6.37±0.01s     0.99  ensemble.RandomForestClassifierBenchmark.time_fit('sparse', 1)
        131±0.8ms          131±1ms     1.00  ensemble.RandomForestClassifierBenchmark.time_predict('dense', 1)
          871±2ms          863±3ms     0.99  ensemble.RandomForestClassifierBenchmark.time_predict('sparse', 1)
  0.7464271763500541  0.7464271763500541     1.00  ensemble.RandomForestClassifierBenchmark.track_test_score('dense', 1)
  0.8656423941766682  0.8656423941766682     1.00  ensemble.RandomForestClassifierBenchmark.track_test_score('sparse', 1)
  0.9968171694224932  0.9968171694224932     1.00  ensemble.RandomForestClassifierBenchmark.track_train_score('dense', 1)
  0.9996123288718864  0.9996123288718864     1.00  ensemble.RandomForestClassifierBenchmark.track_train_score('sparse', 1)
```

on a Linux machine which shows no performance regression.

adam2392 committed:
Revert asv

adam2392 committed:
Merge branch 'main' into crit

jjerphan commented on sklearn/tree/_criterion.pyx:
```diff
@@ -192,6 +180,46 @@ cdef class Criterion:
+    Criterion is the base class for criteria used in supervised tree-based models
+    with a homogeneous float64-dtyped y.
+    """
+    cdef int init(self, const DOUBLE_t[:, ::1] y, DOUBLE_t* sample_weight,
+                  double weighted_n_samples, SIZE_t* samples, SIZE_t start,
+                  SIZE_t end) nogil except -1:
```
```suggestion
    cdef int init(
        self,
        const DOUBLE_t[:, ::1] y,
        DOUBLE_t* sample_weight,
        double weighted_n_samples,
        SIZE_t* samples,
        SIZE_t start,
        SIZE_t end,
    ) nogil except -1:
```

jjerphan commented on sklearn/tree/_criterion.pxd:
```diff
@@ -52,10 +44,27 @@ cdef class Criterion:
+cdef class Criterion(BaseCriterion):
+    """Abstract interface for supervised impurity criteria."""
+
+    # Internal structures
+    cdef const DOUBLE_t[:, ::1] y        # Values of y
```
I think those comments aren't really given extra informations about this instance attribute.
```suggestion
    cdef const DOUBLE_t[:, ::1] y
```

jjerphan commented on sklearn/tree/_criterion.pxd:
```diff
@@ -52,10 +44,27 @@ cdef class Criterion:
     cdef double impurity_improvement(self, double impurity_parent,
-                                     double impurity_left,
-                                     double impurity_right) nogil
+                                    double impurity_left,
+                                    double impurity_right) nogil
```
This change must be reverted.
```suggestion
                                     double impurity_left,
                                     double impurity_right) nogil
```

jjerphan commented on sklearn/tree/_criterion.pxd:
```diff
@@ -52,10 +44,27 @@ cdef class Criterion:
+    # Internal structures
+    cdef const DOUBLE_t[:, ::1] y        # Values of y
+
+    # Methods
```
Same remark regarding this comment.
```suggestion
```

jjerphan reviewed (dismissed):
LGTM, thank you for initiating this work @adam2392 and @jshinm!

Here are a few last comments.

Since this change cornerstone implementations of scikit-learn, I would wait for at least two more approvals from other maintainers before merging this PR.

adam2392 committed:
Apply suggestions from code review

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

adam2392 committed:
Update sklearn/tree/_criterion.pxd

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

adam2392 committed:
Update sklearn/tree/_criterion.pyx

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

adam2392 committed:
Merge branch 'main' into crit

adam2392 commented:
Thanks @jjerphan! Perhaps @thomasjpfan or @glemaitre is interested in giving this a quick look? :)

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Rebased with main

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Merge branch 'main' into crit

adam2392 commented:
> @adam2392 could you solve the conflict since we merge the other PRs.

Kay done!

adam2392 commented:
It does not look like the CI failures are due to this PR?

jjerphan commented:
This is likely a random failure for the Windows' job. Can you push an empty commit or merge `main` in this branch and push?

glemaitre commented:
It is not linked to this PR if you check the log file. No `conda.exe` in the executable `conda` :)

We will see if this is going to be fixed on its own or check that this is reported upstream.

adam2392 committed:
Merge branch 'main' into crit

adam2392 commented:
Not sure why, but it looks like it is failing on main too? I cannot replicate this even locally cuz it works on my Mac at least...

https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=49531&view=logs&j=0238e32a-2fbb-5be1-f782-cfff4ef2924e&t=f063c7d0-643d-578d-31b3-bc7abb593dec

jshinm commented:
could #25091 be related PR?

jjerphan commented:
Yes, it this case merging `main` in your branch and pushing should fix the problem.

jshinm commented:
Thank you for your guidance @jjerphan! Merging `main` as suggested [[here](https://github.com/neurodata/scikit-learn/pull/29)]

jjerphan commented:
You should be able to merge without doing a PR, i.e. by:

```
git checkout <the_branch_you_work_on>
git fetch upstream
git merge main/upstream
git push <your_fork> <the_branch_you_work_on>
```

jshinm committed:
Merge branch 'scikit-learn:main' into crit

jshinm commented:
Thank you @jjerphan! I merged `main` as you suggested 😄

adam2392 commented:
@jshinm seems like we need to wait for someone to review before merging. 

Let's start on the base splitter PR by basing a new branch from this PR

Git checkout crit 
Git checkout -b split
# start making the BaseSplitter PR

adam2392 committed:
Merge branch 'main' into crit

adam2392 commented:
This is ready to go, but if the devs prefer waiting, they can see our proposal for the analagous refactoring in Splitter?

jjerphan commented:
Yes, I am planning to have a look at it.

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Update sklearn/tree/_criterion.pyx

adam2392 commented on sklearn/tree/_criterion.pyx:
```diff
@@ -29,52 +31,31 @@ from ._utils cimport WeightedMedianCalculator
+    scikit-learn's Cython code for criteria.
+
+    The downstream classes _must_ implement methods to compute the impurity
+    in current node and in children nodes.
```
See: https://github.com/scikit-learn/scikit-learn/pull/25118#discussion_r1041297751

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Merge branch 'main' into crit

adam2392 commented:
Just wanted to follow up to get some idea what might be blocking this?

We have several downstream PRs that fully flush out the modularity of criterion/splitter in the right directions w/o introducing performance degradation, so don't wanna continue working on those if something is drastically going to change here.

Thanks!

jjerphan commented:
> Just wanted to follow up to get some idea what might be blocking this?

It has not got response nor attention mainly because of maintainers' lack of time, unfortunately.

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Merge branch 'main' into crit

adam2392 committed:
Merge branch 'main' into crit

jjerphan reviewed (commented):
@Vincent-Maladiere and I are currently taking some time to get a better understanding of those implementations.

In the meantime, I think we better use a feature branch for such refactor (this will allow us to integrate this work more safely and confidently). In this regards, I have created [`refactor/cython-tree-class-hierarchy`](https://github.com/scikit-learn/scikit-learn/tree/refactor/cython-tree-class-hierarchy) from `main`.

@adam2392: could you change the target branch of this PR to `refactor/cython-tree-class-hierarchy`?

adam2392 commented:
> @adam2392: could you change the target branch of this PR to `refactor/cython-tree-class-hierarchy`?

Done!

adam2392 commented:
The only CI issue is the CHANGELOG, but IIRC, that is fine.

adam2392 commented:
Here is an example of how such a refactoring is utilized: https://github.com/neurodata/scikit-tree/blob/main/sktree/tree/_unsup_criterion.pyx

haiatn commented:
Good job!

lorentzenchr commented:
What happened here? Why was it closed?

jjerphan commented:
Sorry for the annoyance.

@adrinjalali mentioned the `scikit-learn:refactor/cython-tree-class-hierarchy` branch to me which seemed not to be used. I did not remember that this PR target to it, hence I deleted it, closing this PR. :facepalm:

I just have repushed the branch since I had a copy on my clone locally.

@adam2392: I cannot reopen this PR, can you?

adam2392 commented:
Sure I can re-open this PR. May I ask what is the desire in doing so? Do I target the now deleted branch, or do I target `main`?

jjerphan commented:
> Do we think this might move forward?

I can't guarantee this will move forward. We need to have people be involved on the Cython side of the project (I personally am contributing to scikit-learn on my free time solely).

[`refactor/cython-tree-class-hierarchy`](https://github.com/scikit-learn/scikit-learn/tree/refactor/cython-tree-class-hierarchy) is available again. Can you target it if you want to continue this PR?
